cpp/src/arrow/parquet/writer.cc

```diff
@@ -143,6 +153,82 @@ Status FileWriter::Close() {
+  const int64_t buffer_offset = chunk * chunk_size * array->type()->value_size();
+  const int64_t value_size = size * array->type()->value_size();
+  auto chunk_buffer = std::make_shared<Buffer>(array->data(), buffer_offset, value_size);
+  std::shared_ptr<Buffer> null_bitmap;
+  int32_t null_count = 0;
+  if (array->null_count() > 0) {
+    int64_t null_offset = (chunk * chunk_size) / 8;
+    int64_t null_size = util::ceil_byte(size) / 8;
+    null_bitmap = std::make_shared<Buffer>(array->null_bitmap(), null_offset, null_size);
+    for (int64_t k = 0; k < size; k++) {
+      if (!util::get_bit(null_bitmap->data(), k)) { null_count++; }
+    }
+  }
+  std::shared_ptr<Array> out_array;
+  RETURN_NOT_OK(MakePrimitiveArray(
+      array->type(), size, chunk_buffer, null_count, null_bitmap, &out_array));
+  *out = std::static_pointer_cast<PrimitiveArray>(out_array);
+  return Status::OK();
+}
+
+Status WriteFlatTable(const Table* table, MemoryPool* pool,
+    std::shared_ptr<::parquet::OutputStream> sink, int64_t chunk_size) {
+  // Ensure alignment of sliced PrimitiveArray, esp. the null bitmap
+  // TODO: Support other chunksizes than multiples of 512
+  if (((chunk_size & 511) != 0) && (chunk_size != table->num_rows())) {
+    return Status::NotImplemented(
+        "Only chunk sizes that are a multiple of 512 are supported");
+  }
+
+  std::shared_ptr<::parquet::SchemaDescriptor> parquet_schema;
+  RETURN_NOT_OK(ToParquetSchema(table->schema().get(), &parquet_schema));
+  auto schema_node = std::static_pointer_cast<GroupNode>(parquet_schema->schema());
+  std::unique_ptr<ParquetFileWriter> parquet_writer =
+      ParquetFileWriter::Open(sink, schema_node);
+  FileWriter writer(pool, std::move(parquet_writer));
+
+  // TODO: Support writing chunked arrays.
+  for (int i = 0; i < table->num_columns(); i++) {
+    if (table->column(i)->data()->num_chunks() != 1) {
+      return Status::NotImplemented("No support for writing chunked arrays yet.");
+    }
+  }
+
+  // Cast to PrimitiveArray instances as we work with them.
+  std::vector<std::shared_ptr<PrimitiveArray>> arrays(table->num_columns());
+  for (int i = 0; i < table->num_columns(); i++) {
+    // num_chunks == 1 as per above loop
+    std::shared_ptr<Array> array = table->column(i)->data()->chunk(0);
+    auto primitive_array = std::dynamic_pointer_cast<PrimitiveArray>(array);
+    if (!primitive_array) {
+      return Status::NotImplemented("Table must consist of PrimitiveArray instances");
+    }
+    arrays[i] = primitive_array;
+  }
+
+  for (int chunk = 0; chunk * chunk_size < table->num_rows(); chunk++) {
+    int64_t size = std::min(chunk_size, table->num_rows() - chunk * chunk_size);
+    RETURN_NOT_OK(writer.NewRowGroup(size));
+    for (int i = 0; i < table->num_columns(); i++) {
+      std::shared_ptr<PrimitiveArray> array;
+      RETURN_NOT_OK(TemporaryArraySlice(chunk, chunk_size, arrays[i].get(), &array));
+      RETURN_NOT_OK(writer.WriteFlatColumnChunk(array.get()));
```

Comment:
I wonder if this might be simpler if you pass an offset and size here rather than having to produce a temporary slice